apps/fyler/src/fyler_category_manager.erl

```diff
@@ -0,0 +1,694 @@
+
+code_change(_OldVsn, StateName, State, _Extra) ->
+  {ok, StateName, State}.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+apply_state(start, Event, State) ->
+  start(Event, State);
+
+apply_state(zero, Event, State) ->
+  zero(Event, State);
+
+apply_state(pools, Event, State) ->
+  pools(Event, State);
+
+apply_state(tasks, Event, State) ->
+  tasks(Event, State).
+
+%% @doc Choose pool
+
+choose_pool(#state{retasks = ReTasks, tasks = Tasks, pools = Pools} = State) ->
+  {NextState, [NewReTasks, NewTasks], NewPools} = choose_pool([ReTasks, Tasks], Pools, [], []),
+  {NextState, State#state{retasks = NewReTasks, tasks = NewTasks, pools = NewPools}}.
+
+choose_pool([Queue|Queues], [{Node, enabled}|Pools], QueueAcc, PoolsAcc) ->
+  case fyler_queue:out(Queue) of
+    {empty, Queue} ->
+      choose_pool(Queues, [{Node, enabled}|Pools], [Queue|QueueAcc], PoolsAcc);
+    {{value, Task}, NewQueue} ->
+      case send_to_pool(Node, Task) of
+        {Ref, Id} -> choose_pool([NewQueue|Queues], Pools, QueueAcc, [{Node, {pending, Ref, Id}}|PoolsAcc]);
+        _ -> choose_pool([NewQueue|Queues], [{Node, enabled}|Pools], QueueAcc, PoolsAcc)
+      end
+  end;
+
+choose_pool([Queue|Queues], [{Node, Status}|Pools], QueueAcc, PoolsAcc) ->
+  choose_pool([Queue|Queues], Pools, QueueAcc, [{Node, Status}|PoolsAcc]);
+
+choose_pool([], Pools, QueueAcc, PoolsAcc) ->
+  {pools, lists:reverse(QueueAcc), Pools ++ lists:reverse(PoolsAcc)};
+
+choose_pool(Queues, [], QueueAcc, PoolsAcc) ->
+  NewState =
+    case lists:any(fun(Q) -> not fyler_queue:is_empty(Q) end, Queues) of
+      false ->
+        zero;
+      true ->
+        tasks
+    end,
+  {NewState, lists:reverse(QueueAcc) ++ Queues, lists:reverse(PoolsAcc)}.
+
+%% @doc Get task from ets by id
+
+task(Category, Id) ->
+  case ets:lookup(Category, Id) of
+    [#current_task{task = Task}] ->
+      Task;
+    _ ->
+      undefined
+  end.
+
+%% @doc Restart tasks
+
+restart_tasks(Category, Queue, Node) ->
+  Tasks = ets:match_object(Category, #current_task{pool = Node, _= '_'}),
+  In = fun(#current_task{task = Task}, Q) -> fyler_queue:in(Task, Q) end,
+  lists:foldl(In, Queue, Tasks).
+
+%% @doc Send task to remote pool
+
+send_to_pool(Node, Task = #task{category = Category, id = Id}) ->
+  case task(Category, Id) of
+    undefined ->
+      undefined;
+    _ ->
+      update_ets(Node, Task),
+      Ref = make_ref(),
+      rpc:cast(Node, fyler_pool, run_task, [Task, Ref]),
+      {Ref, Id}
+  end.
+
+update_ets(Task) ->
+  update_ets(undefined, Task).
+
+update_ets(Node, #task{id = Id, type = Type, category = Category, file = #file{url = Url}} = Task) ->
+  Status =
+      case Node of
+          undefined -> queued;
+          _ -> progress
+      end,
+  ets:insert(Category, #current_task{id = Id, task = Task, type = Type, url = Url, pool = Node, status = Status}).
```

Comment:
Line 555 is too long:   ets:insert(Category, #current_task{id = Id, task = Task, type = Type, url = Url, pool = Node, status = Status})..